include/seqan3/core/debug_stream/default_printer.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <iosfwd>
+#include <utility>
+#include <tuple>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+struct no_printer_found{};
+template <typename> struct advanceable_alignment_coordinate_printer {};
+template <typename> struct alignment_matrix_printer {};
+template <typename> struct alignment_printer {};
+template <typename> struct alignment_result_printer {};
+template <typename> struct alphabet_printer {};
+template <typename> struct debug_stream_printer {};
+template <typename> struct input_range_printer {};
+template <typename> struct integer_sequence_printer {};
+template <typename> struct integral_printer {};
+template <typename> struct mask_printer {};
+template <typename> struct optional_printer {};
+template <typename> struct sequence_printer {};
+template <typename> struct std_printer {};
+template <typename> struct char_sequence_printer {};
+template <typename> struct trace_directions_printer {};
+template <typename> struct tuple_printer {};
```

Comment:
Having all forward declarations in one place is nice, since you know what overloads we have.

The complete mechanism is based on [two-phase_lookup](https://en.cppreference.com/w/cpp/language/two-phase_lookup)